thrust/testing/tabulate_output_iterator.cu

```diff
@@ -0,0 +1,150 @@
+#include <thrust/copy.h>
+#include <thrust/device_vector.h>
+#include <thrust/functional.h>
+#include <thrust/gather.h>
+#include <thrust/host_vector.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/tabulate_output_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/reduce.h>
+#include <thrust/sequence.h>
+
+#include <cuda/std/type_traits>
+
+#include <unittest/unittest.h>
+
+template <typename OutItT>
+struct host_write_op
+{
+  OutItT out;
+
+  template <typename IndexT, typename T>
+  _CCCL_HOST void operator()(IndexT index, T val)
+  {
+    out[index] = val;
+  }
+};
```

Comment:
Remark: Looking at you storing a secondary iterator for output in the callable, I think I would have designed the tabulate iterator differently, or maybe extended your version:

If the tabulate iterator would store another iterator underneath it, and move it along, you would not require random access to `out` here when writing to the tabulate iterator. Your function would then have a signature like:
```c++
  template <typename IndexT, typename T, typename It>
  _CCCL_HOST void operator()(IndexT index, T val, It underlying_iterator)
  {
    // index unused here, but could be use to alter `val` or verify something
    *underlying_iterator = val;
  }
```
Just an idea.

Reply:
I can understand the argument. It sounds like another specialization that has the advantage of not requiring random access. However, some use cases won't have that _one_ underlying iterator or any underlying iterator at all. At which point I think it may feel a little out of place having to pass in something like a `discard_iterator` there. I would prefer to leave that to follow-on work on a sibling iterator to `tabulate_output_iterator`, wdyt?

Reply:
> I would prefer to leave that to follow-on work

I agree!